### ISSUE TITLE
Inline functions in a referenced project optimized incorrectly

### ISSUE THREAD

**Issue by rca22** [dotnet/fsharp#17903]
In the attached example, we create three types; 
- ID, in ClassLibrary1
- ID2, in ClassLibrary1, which relies on ID
- ID3 in ConsoleApp1, which relies on ID2.

In each of these we override equality and comparison operators, using inline functions to improve performance for ID3. (Note that this repro is a minimal example from more complicated code where this makes more sense!)

We create instances of each in the console app. and check that two unequal values of each type are reported as unequal by the compiler.
In Release mode only, the third assertion, that two obviously different instances of ID3 are unequal fails.

To repro, please run the attached program.
[repro.zip](https://github.com/user-attachments/files/17460199/repro.zip)

**Expected behavior**

I expect the program to exit with code 0.

**Actual behavior**

An exception is thrown in the last equality check, for `aBarBar` versus `bBarBar`

**Known workarounds**

Running this in debug mode fixes the issue.
Putting the definitions of ID and ID2 in the same project as ID3 fixes the issue. Because of this, I suspect this issue is something to do with incorrect optimisations around inlining functions from other projects.
Changing ID from a struct into a class by removing all the attributes associated with it fixes the issue.

**Related information**

Tested in .NET 8, SDKs 8.0.100 and 8.0.403
Windows 10.
VS 2022 for editing.

**Comment by vzarytovskii** [comment]
> An exception is thrown in the last equality check, for aBarBar versus bBarBar

What's the exception? What's yielded in the nightly 9.0 sdk?

**Comment by rca22** [comment]
> > An exception is thrown in the last equality check, for aBarBar versus bBarBar
> 
> What's the exception? What's yielded in the nightly 9.0 sdk? Does it change with realsig off? It could also be related to the additional equality generated.

To be clear the last equality check reads `if aBarBar = bBarBar then failwith "Expect aBarBar and bBarBar to differ"`; earlier `aBarBar` and `bBarBar` are set up to be clearly different. The exeption is not an internal one; just there to indicate the deviation from expected behaviour.

**Comment by vzarytovskii** [comment]
> > > An exception is thrown in the last equality check, for aBarBar versus bBarBar
> > 
> > What's the exception? What's yielded in the nightly 9.0 sdk? Does it change with realsig off? It could also be related to the additional equality generated.
> 
> To be clear the last equality check reads `if aBarBar = bBarBar then failwith "Expect aBarBar and bBarBar to differ"`; earlier `aBarBar` and `bBarBar` are set up to be clearly different. The exception is not an internal one; just there to indicate the deviation from expected behaviour.

I see, thanks. It appears that it's failing in SDK 6.0, could've been there forever.

**Comment by T-Gro** [comment]
This is what the decompiled Release code looks like for the Equals check of ID3.
As you can see, the name ID4 was being reused for both sides, ending up comparing a string with the same string, all the time.

![Image](https://github.com/user-attachments/assets/58790bf2-44f9-4bdc-bfae-2edaa0b7a241)

**Comment by vzarytovskii** [comment]
> This is what the decompiled Release code looks like for the Equals check of ID3.
> As you can see, the name ID4 was being reused for both sides, ending up comparing a string with the same string, all the time.
> 
> ![Image](https://github.com/user-attachments/assets/58790bf2-44f9-4bdc-bfae-2edaa0b7a241)
> 

That's some awkward codegen. We will need to see whether it's a recent regression. If it is, we need to fix it for 9.0.200

**Comment by vzarytovskii** [comment]
This is minimal repro:

Project1: Library/Library.fs
```fsharp
namespace Library

open System

#nowarn "345"
#nowarn "346"

[<Struct; NoComparison; CustomEquality>]
type ID (value : string) =

    member _.Value = value

    member inline this.FastEquals(other: ID) =
        this.Value = other.Value

    override this.Equals(other : obj) =
        match other with
        | :? ID as other -> this.FastEquals other
        | _ -> false


[<Struct; NoComparison; CustomEquality>]
type ID2 =
    {
        Value : ID
    }

    member inline this.FastEquals(other: ID2) =
        this.Value.FastEquals other.Value

    override this.Equals(other : obj) =
        match other with
        | :? ID2 as other -> this.FastEquals other
        | _ -> false
```
Project2: Console/Program.fs:
```fsharp

**Comment by vzarytovskii** [comment]
Updated minimal repro:

Library.fs:
```fsharp
namespace Library

#nowarn "345"
#nowarn "346"

[<Struct; NoComparison; CustomEquality>]
type ID (value : string) =
    member _.Value = value
    member inline this.FastEquals(other: ID) = this.Value = other.Value
    override _.Equals(_ : obj) = failwith "Not needed"

[<NoComparison; CustomEquality>]
type ID2 =
    {
        Value : ID
    }

    member inline this.FastEquals(other: ID2) = this.Value.FastEquals other.Value

    override _.Equals(_ : obj) = failwith "Not needed"

```

Program.fs:

```fsharp
open Library

[<EntryPoint>]
let main _ =

    let aBar = { Value = ID "a" }
    let bBar = { Value = ID "b" }


    if aBar.FastEquals(bBar) then failwith "1: Expect aBarBar and bBarBar to differ"

    0
```

**Comment by vzarytovskii** [comment]
Update:
It is also not the case with static members. So the issue is that local seems not to be introduced for the `this` parameter of the struct member.

**Comment by vzarytovskii** [comment]
TLDR of the issue is that we reuse locals when we shouldn't be .

**Comment by vzarytovskii** [comment]
So it seems that we don't mark `copyOfStruct` (copy of `this`) as used local. It seems it's in the wrong place in the TAST.

**Comment by vzarytovskii** [comment]
So, the issue is that we leak address of this defensive copy out of its scope, one immediate solution would be is to never realloc for `copyOfStruct` (it's a special compiler name, we don't indicate it any other way):

```
showTerm: pass-end:
top implementation 
  Module Defs
    open ... 
    Module Defs
      module Program
        Module Defs
          open ... 
          let main [<.ctor>] ! <0>[1] =
                fun _arg1 ->
                  let aBar =
                        {Value=.ctor "a"}
                  let bBar =
                        {Value=.ctor "b"}
                  __debugPoint( /Users/u/code/issues/gh/17903/ConsoleApp1/Program.fs(9,4)-(9,20) )
                  let this =
                        let mutable copyOfStruct =
                              aBar.# Value
                        & copyOfStruct ()
                  let otherid =
                        bBar.# Value
                  System.Console.WriteLine [System.String.Concat [get_Value this ();" ";get_Value (& otherid ()) ()]]
                  __debugPoint( /Users/u/code/issues/gh/17903/ConsoleApp1/Program.fs(11,4)-(11,5) )
                  0
```

**Comment by vzarytovskii** [comment]
We shouldn't be generating it in general, but for now not reallocating it is fine:
![Image](https://github.com/user-attachments/assets/080ed7bc-77e3-438a-8588-ec5667e53525)
